keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1655 @@
+   // the node for which it wants the resource to be allocated
+   // before the driver proceeds with allocation.
+   //
+   // For immediate allocation, the scheduler will not set
+   // this field. The resource driver controller may
+   // set it to trigger allocation on a specific node if the
+   // resources are local to nodes.
+   SelectedNode string
+
+   // When allocation is delayed, and the scheduler needs to
+   // decide on which node a Pod should run, it will first
+   // ask the driver on which nodes the resource might be
+   // made available. To trigger that check, the scheduler
+   // provides the names of nodes which might be suitable
+   // for the Pod. Will be updated periodically until
+   // the claim is allocated.
+   PotentialNodes []string
+
+   // A change of the PotentialNodes field triggers a check
+   // in the driver on which of those nodes the resource
+   // might be made available. It then excludes nodes
+   // by listing those where that is not the case in
+   // UnsuitableNodes.
+   //
+   // Unsuitable nodes will be ignored by the scheduler
+   // when selecting a node for a Pod. All other nodes are
+   // potential candidates, either because no information
+   // is available yet or because allocation might
+   // succeed.
+   //
+   // This can change, so the driver must refresh
+   // this information periodically and/or after
+   // changing resource allocation for some other
+   // ResourceClaim until a node gets selected by
+   // the scheduler.
+   UnsuitableNodes []string
+
+   // This field will get set by the resource driver to
+   // inform the scheduler where it can schedule Pods
+   // using the ResourceClaim.
+   //
+   // A resource driver may already set this before
+   // the resource is allocated. The scheduler will
+   // then check this field in addition to UnsuitableNodes
+   // to filter out nodes where the resource cannot
+   // be allocated.
+   //
+   // For an allocated resource this field defines
+   // where the resource can be used by Pods.
+   //
+   // Setting this field is optional. If nil, the
+   // resource is available everywhere.
+   AvailableOnNodes *corev1.NodeSelector
+
+   // Arbitrary data returned by the driver after a successful allocation.
+   // This data is passed to the driver for all operations involving
+   // the allocated resource. This is opaque for Kubernetes.
+   // Driver documentation may explain to users how to interpret
+   // this data if needed.
+   //
+   // The attributes must be sufficient to deallocate the resource
+   // because the ResourceClass might not be available anymore
+   // when deallocation starts.
+   Attributes map[string]string
+
+    // UsersLimit determines how many entities are allowed to use this resource
+    // at the same time. The default is 1. -1 enables the usage by an unlimited number
+   // of users. Individual containers in a pod are not counted as users, only the Pod
+   // is.
+   UserLimit int
```

Comment:
this name is confusing.

is there any tenancy issue that needs to be accounted for?

for example, pods in namespace A may not share devices with pods in namespace B.

Reply:
No, this isn't users in the sense of "users of the cluster". Multi-tenancy is handled by limiting ResourceClaims to individual namespaces, but that's not what this is about.

Perhaps `MaximumReservationCount` would be a better name?

Reply:
I had suggested the following previously: https://github.com/kubernetes/enhancements/pull/3064#discussion_r794433346

i.e. Maybe `MaxReservations` to match the `ReservedFor` field below?